### PR TITLE
feat(medusa): Introduce Payment Processor API

### PR DESCRIPTION
**Pull request by adrien2p**
**What**

This PR only provides a proposal for the payment service plugin interface and abstract class. It purposely does not provide any other changes in order to first agreed on the API we want to provide before proceeding toward any changes.

**NOTE**

Depending on the decision to take regarding the support of both API. This proposal might slightly change

FIXES CORE-886

### CODE REVIEW

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (olivermrbl) [inline packages/medusa/src/interfaces/payment-processor.ts:21]: thought: I don't think we need this method in the interface. At least not in the first iteration.

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (olivermrbl) [inline packages/medusa/src/interfaces/payment-processor.ts:1]: suggestion: Would change `sessionId` in all methods to a more generic variable name such as `paymentId`.  Also, for refund, capture, etc. session id is misleading as we are actually updating the payment.

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (olivermrbl) [inline packages/medusa/src/interfaces/payment-processor.ts:26]: suggestion: Rename to `initiatePayment(... ` to prevent it from getting mixed up with our own `createPayment` of our `PaymentProviderService`

File: packages/medusa/src/types/payment-processor.ts
Comment (olivermrbl) [inline packages/medusa/src/types/payment-processor.ts:8]: suggestion: Add support for `billing_address`

File: packages/medusa/src/types/payment-processor.ts
Comment (olivermrbl) [inline packages/medusa/src/types/payment-processor.ts:14]: question: Do you think we should allow for `email` and `[shipping | billing]_address` without a cart? Potentially also `context`

File: packages/medusa/src/types/payment-processor.ts
Comment (adrien2p) [reply packages/medusa/src/types/payment-processor.ts:8]: I was thinking, do we really need the cart data for the payment processor? Or would it be more for the checkout API?

File: packages/medusa/src/types/payment-processor.ts
Comment (adrien2p) [reply packages/medusa/src/types/payment-processor.ts:14]: ahah, partly part of my previous comment

File: packages/medusa/src/types/payment-processor.ts
Comment (olivermrbl) [reply packages/medusa/src/types/payment-processor.ts:8]: No, as this will be an entirely new interface, I think it would be best to remove the need for a cart in the context of payments and instead rely on a set of required props

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-processor.ts:21]: Yes I can remove it, it was more to move the initialisation of the provider from the constructor to this method knowing that it can be async. I ll remove it

Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-processor.ts:21]: @olivermrbl thoughts, the initialization of an external package can be async. Since the constructor is always since it is possible to call a method while the initialization is not finalized cause the promise is not resolved yet. wdyt?

Comment (olivermrbl) [reply packages/medusa/src/interfaces/payment-processor.ts]: I am fine keeping it in. But then I suggest renaming it to be specific to the initialization of a given SDK/client/etc. and the current `init` could be mixed up with the initialization of a payment. Wdyt?

Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-processor.ts]: `initialisePaymentProcessor` ? wdyt?

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (olivermrbl) [inline packages/medusa/src/interfaces/payment-processor.ts:42]: thought: Sorry, I actually liked init for the SDK/client/etc better than initializePayment, so maybe we could stick with that and then rename this one to startPayment. Wdyt? 😅

Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-processor.ts:42]: InitiatePayment was previously createPayment. I ve removed the init method. Do you want me to re add the init? I am not sure i ve followed

Comment (olivermrbl) [reply packages/medusa/src/interfaces/payment-processor.ts:42]: Sorry, I misplaced my comment. To avoid blocking this PR further, I'm fine with just using `init`. My apologies for going back and forth on this one.

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (olivermrbl) [reply packages/medusa/src/interfaces/payment-processor.ts]: Sorry, I misplaced my comment. To avoid blocking this PR further, I'm fine with just using `init`. My apologies for going back and forth on this one.

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (olivermrbl) [inline packages/medusa/src/interfaces/payment-processor.ts:11]: question(non-blocking): What did you think of adding `billing_address`?

Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-processor.ts:11]: oh sorry I must have missed it, I ll add it straight away

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (olivermrbl) [inline packages/medusa/src/interfaces/payment-processor.ts:9]: question(non-blocking): Would it make sense to include a `context` or `details` object to store information specific to the provider being used?

Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-processor.ts:9]: what information are you thinking of?

File: .changeset/thirty-parents-design.md
Comment (olivermrbl) [inline .changeset/thirty-parents-design.md:5]:
```suggestion
feat(medusa): Payment Processor API
```

**Review by olivermrbl — APPROVED** [review]
LGTM! Pending @srindom's approval :)

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (srindom) [reply packages/medusa/src/interfaces/payment-processor.ts:9]: An example is Adyen which in some instances need the customer IP to perform fraud checks; having an open ended prop that can take processor specific data like this can be helpful.

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (srindom) [inline packages/medusa/src/interfaces/payment-processor.ts:11]: Both ShippingAddress and ShippingMethods seem out of place here. We want this to be only for processing payments of an amount in a given currency. 

Think it will be Checkout Provider API that would use things like Shipping details. Wdyt?

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (srindom) [inline packages/medusa/src/interfaces/payment-processor.ts:38]: thought: A scenario came up the other day where a user needed to register two of the same plugins to their store. Specifically, they had two PayPal accounts: 1 for EUR and 1 for USD. It would be valuable if we could open up for customizing the identifier through `medusa-config.js`'s plugin options. 

This probably has more to do with how we register plugins.

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (srindom) [inline packages/medusa/src/interfaces/payment-processor.ts:43]: question: at what point in the life cycle does this get called?

Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-processor.ts:43]: I thought calling it as part of either the loader or the install provider. wdyt?

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-processor.ts:38]: yes got it,

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-processor.ts:9]: I will add it indeed, how is it managed at the moment?

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-processor.ts:38]: I would have assumed that this is something managed by the Paypal API depending on the currency that is used. Don't you think it would need a bit more investigation?

**Comment by adrien2p** [comment]
@olivermrbl @srindom any other points to look at for that pr?

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (olivermrbl) [reply packages/medusa/src/interfaces/payment-processor.ts:33]: I think @srindom's idea is excellent, but I suggest we keep it for a separate ticket to not block this PR further, as we would prob. need to rethink payment plugin registration.

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (olivermrbl) [reply packages/medusa/src/interfaces/payment-processor.ts:38]: As this would also need to be added to the payment plugin registration, I would keep it for a separate PR. Perhaps the one where we address the other topic.

So should we leave it out for now?

**Review by olivermrbl — DISMISSED** [review]
LGTM, pending @srindom's approval :)

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-processor.ts:38]: Here the purpose was to describe the API. everything related to the loader etc will be made in the next pr @olivermrbl

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (adrien2p) [inline packages/medusa/src/interfaces/payment-processor.ts:53]: @olivermrbl @srindom I was working on the next step (loader and payment provider support) and I came to think that we have no idea of what is the paymentId specifically. In fact, we are saving the data in the payment session data and only the plugin is aware of where is the payment id located and what is the name of the property. For stripe it is the `id` from the session data, but it seems wrong to assume that it is always the case. What do you think if instead, the first argument is the payment session data?

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (srindom) [inline packages/medusa/src/interfaces/payment-processor.ts:138]: thought: think we should kill this in the first iteration.

**Review by srindom — APPROVED** [review]
LGTM - would suggest we remove `getSavedMethods` but otherwise ready to merge from my POV

File: packages/medusa/src/interfaces/payment-processor.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-processor.ts:138]: Yeah probably, it means that the get-payment-method end point will not be effective for a payment processor. Is that ok?

**Pull request merged**
